pkg/chunk/cassandra/storage_client.go

```diff
@@ -0,0 +1,219 @@
+	cluster.QueryObserver = observer{}
+
+	return cluster.CreateSession()
+}
+
+// createKeyspace will create the desired keyspace if it doesn't exist.
+func (cfg *Config) createKeyspace() error {
+	cluster := gocql.NewCluster(cfg.address)
+	cluster.Keyspace = "system"
+	cluster.Timeout = 20 * time.Second
+	session, err := cluster.CreateSession()
+	if err != nil {
+		return err
+	}
+	defer session.Close()
+
+	return session.Query(fmt.Sprintf(
+		`CREATE KEYSPACE IF NOT EXISTS %s
+		 WITH replication = {
+			 'class' : 'SimpleStrategy',
+			 'replication_factor' : %d
+		 }`,
+		cfg.keyspace, cfg.replicationFactor)).Exec()
+}
+
+// storageClient implements chunk.storageClient for GCP.
+type storageClient struct {
+	cfg       Config
+	schemaCfg chunk.SchemaConfig
+	session   *gocql.Session
+}
+
+// NewStorageClient returns a new StorageClient.
+func NewStorageClient(cfg Config, schemaCfg chunk.SchemaConfig) (chunk.StorageClient, error) {
+	session, err := cfg.session()
+	if err != nil {
+		return nil, err
+	}
+
+	return &storageClient{
+		cfg:       cfg,
+		schemaCfg: schemaCfg,
+		session:   session,
+	}, nil
+}
+
+func (s *storageClient) Close() {
+	s.session.Close()
+}
+
+type writeBatch struct {
+	b *gocql.Batch
+}
+
+func (s *storageClient) NewWriteBatch() chunk.WriteBatch {
+	return writeBatch{
+		b: gocql.NewBatch(gocql.UnloggedBatch),
+	}
+}
+
+func (b writeBatch) Add(tableName, hashValue string, rangeValue []byte, value []byte) {
+	b.b.Query(fmt.Sprintf("INSERT INTO %s (hash, range, value) VALUES (?, ?, ?)", tableName),
+		hashValue, rangeValue, value)
+}
+
+func (s *storageClient) BatchWrite(ctx context.Context, batch chunk.WriteBatch) error {
+	cassandraBatch := batch.(writeBatch)
+	return s.session.ExecuteBatch(cassandraBatch.b.WithContext(ctx))
+}
+
+func (s *storageClient) QueryPages(ctx context.Context, query chunk.IndexQuery, callback func(result chunk.ReadBatch, lastPage bool) (shouldContinue bool)) error {
+	var q *gocql.Query
+	if len(query.RangeValuePrefix) > 0 {
+		q = s.session.Query(fmt.Sprintf("SELECT range, value FROM %s WHERE hash = ? AND range >= ?", query.TableName),
+			query.HashValue, query.RangeValuePrefix)
+	} else if len(query.RangeValueStart) > 0 {
+		q = s.session.Query(fmt.Sprintf("SELECT range, value FROM %s WHERE hash = ? AND range >= ?", query.TableName),
+			query.HashValue, query.RangeValueStart)
+	} else {
+		q = s.session.Query(fmt.Sprintf("SELECT range, value FROM %s WHERE hash = ?", query.TableName),
+			query.HashValue)
+	}
+
+	iter := q.WithContext(ctx).Iter()
+	defer iter.Close()
+	scanner := iter.Scanner()
+	for scanner.Next() {
+		var b readBatch
+		if err := scanner.Scan(&b.rangeValue, &b.value); err != nil {
+			return err
+		}
+		if callback(b, false) {
+			return nil
+		}
+	}
+	return scanner.Err()
+}
+
+// readBatch represents a batch of rows read from Cassandre.
+type readBatch struct {
+	rangeValue []byte
+	value      []byte
+}
+
+func (readBatch) Len() int {
+	return 1
```

Comment:
Nit: a comment about:

- why this needs to be `1` and 
- why we need this `Len()` method at all
  (because `readBatch` has to implement [this interface](github.com/weaveworks/cortex/blob/HEAD/pkg/chunk/storage_client.go#L25)) 

might help the next reader/maintainer.